src/main/impl/Config.cpp

```diff
@@ -0,0 +1,422 @@
+#include <boost/algorithm/string.hpp>
+
+#include <main/Config.h>
+
+#include <fstream>
+#include <iostream>
+
+boost::json::object
+parseConfig(std::string const& filename)
+{
+    try
+    {
+        std::ifstream in(filename, std::ios::in | std::ios::binary);
+        if (in)
+        {
+            std::stringstream contents;
+            contents << in.rdbuf();
+            in.close();
+            std::cout << contents.str() << std::endl;
+            boost::json::value value = boost::json::parse(contents.str());
+            return value.as_object();
+        }
+    }
+    catch (std::exception const& e)
+    {
+        std::cout << e.what() << std::endl;
+    }
+
+    throw std::runtime_error(
+        "Could not parse JSON configuration. Verify it is valid JSON format");
+}
+
+struct cfgOPTIONAL;
+struct cfgREQUIRED;
+
+template <typename T, typename F>
+void
+add(T& value, boost::json::value& json, F&& parser)
+{
+    try
+    {
+        value = parser(json);
+    }
+    catch (std::runtime_error const& e)
+    {
+        throw std::runtime_error(
+            "Could not parse field " + boost::json::serialize(json) + "\n" +
+            e.what());
+    }
+}
+
+template <typename T, typename F, typename Def>
+void
+add(T& value, boost::json::value& json, F&& parser, Def def)
+{
+    try
+    {
+        add(value, json, std::move(parser));
+    }
+    catch (std::exception const&)
+    {
+        value = def;
+    }
+}
+
+template <typename T, typename F>
+void
+addPolymorphic(
+    T& value,
+    std::unique_ptr<T>& store,
+    boost::json::value& json,
+    F&& parser)
+{
+    add(store, json, parser);
+    value = *store;
+}
+
+template <typename Required, typename T>
+auto
+checkRequired(std::optional<T>& potentiallyRequired)
+{
+    if constexpr (std::is_same<Required, cfgREQUIRED>::value)
+    {
+        if (!potentiallyRequired)
+            throw std::runtime_error("Required value not present");
+
+        return *potentiallyRequired;
+    }
+    else
+    {
+        return potentiallyRequired;
+    }
+}
+
+template <typename Required>
+auto
+parseString(boost::json::value& value)
+{
+    std::optional<std::string> parsed = {};
+
+    if (value.is_string())
+        parsed = value.as_string().c_str();
+
+    return checkRequired<Required>(parsed);
+}
+
+template <typename Required>
```

Comment:
So `Required` is really just a bool? Why do this templating stuff, why not just pass a bool, or have two functions?

Reply:
Because we want `parseUInt<cfgOptional>()` to return a `std::optional<std::uin32_t>` but we want `parseUInt<cfgRequired>()` to return a `std::uint32_t`